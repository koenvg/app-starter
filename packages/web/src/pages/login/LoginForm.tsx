import { StyledForm, LoginButton } from "./LoginForm.styled";
import { useAuthentication } from "../../components/authentication/AuthenticationProvider";
import React, { SFC } from "react";
import { useFormik } from "formik";
import { TextField, Typography } from "@material-ui/core";

export const LoginForm: SFC = () => {
  const { authenticate } = useAuthentication();
  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
      general: "",
    },
    validate: ({ email, password }) => {
      const errors: { email?: string; password?: string } = {};
      if (email === "") {
        errors.email = "Please enter your username";
      }
      if (password === "") {
        errors.password = "Please enter your password";
      }

      return errors;
    },
    onSubmit: async ({ email, password }, { setSubmitting, setErrors }) => {
      try {
        await authenticate(email, password);
      } catch (e) {
        setSubmitting(false);
        setErrors({ general: "Your username and/or password do not match" });
      }
    },
  });

  return (
    <StyledForm noValidate onSubmit={formik.handleSubmit}>
      <TextField
        variant="outlined"
        margin="normal"
        required
        fullWidth
        id="email"
        label="Email Address"
        name="email"
        autoComplete="email"
        autoFocus
        onChange={formik.handleChange}
        value={formik.values.email}
        disabled={formik.isSubmitting}
        error={!!formik.errors.email}
        helperText={formik.errors.email}
      />
      <TextField
        variant="outlined"
        margin="normal"
        required
        fullWidth
        name="password"
        label="Password"
        type="password"
        id="password"
        autoComplete="current-password"
        onChange={formik.handleChange}
        value={formik.values.password}
        disabled={formik.isSubmitting}
        error={!!formik.errors.password}
        helperText={formik.errors.password}
      />
      {formik.errors.general && (
        <Typography variant="body2" color="error">
          {formik.errors.general}
        </Typography>
      )}
      <LoginButton type="submit" fullWidth variant="contained" color="primary" loading={formik.isSubmitting}>
        <Typography>Sign In</Typography>
      </LoginButton>
    </StyledForm>
  );
};

LoginForm.displayName = "LoginForm";
