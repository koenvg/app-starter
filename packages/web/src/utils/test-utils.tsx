import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "components";

const AllTheProviders: React.SFC = ({ children }) => {
  return <ThemeProvider>{children}</ThemeProvider>;
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "queries" | "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
