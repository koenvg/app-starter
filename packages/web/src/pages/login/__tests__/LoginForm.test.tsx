const mockAuthenticate = jest.fn().mockResolvedValue({});

jest.mock("../../../components/authentication/AuthenticationProvider", () => ({
  useAuthentication: () => ({ authenticate: mockAuthenticate }),
}));

import { LoginForm } from "../LoginForm";
import { render } from "../../../utils/test-utils";
import React from "react";
import { wait, fireEvent, findByText } from "@testing-library/react";

describe("LoginForm", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const signIn = async (container: HTMLElement) => {
    const email = container.querySelector('input[name="email"]') as Element;
    const password = container.querySelector('input[name="password"]') as Element;
    const submit = container.querySelector('button[type="submit"]') as Element;

    await wait(() => {
      fireEvent.change(email, {
        target: {
          value: "test@test.com",
        },
      });
    });

    await wait(() => {
      fireEvent.change(password, {
        target: {
          value: "somePassword",
        },
      });
    });

    await wait(() => {
      fireEvent.click(submit);
    });
  };

  it("should render the login form", () => {
    const component = render(<LoginForm />);
    expect(component.container).toMatchSnapshot();
  });

  it("should show an error when the user does not fill in username or password", async () => {
    const component = render(<LoginForm />);

    fireEvent.submit(component.getByText("Sign In"));

    await findByText(component.container, "Please enter your password");
    expect(component.container).toMatchSnapshot();
  });

  it("should sign the user in the they enter username and password", async () => {
    const { container } = render(<LoginForm />);
    await signIn(container);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should show a generic error when one occurs during signing in", async () => {
    mockAuthenticate.mockRejectedValueOnce(new Error("Epic failure"));
    const { container } = render(<LoginForm />);
    await signIn(container);
    expect(container.firstChild).toMatchSnapshot();
  });
});
