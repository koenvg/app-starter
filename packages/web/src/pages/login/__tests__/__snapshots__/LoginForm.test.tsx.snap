// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should render the login form 1`] = `
.c0 {
  width: 100%;
  margin-top: 8;
}

.c1.c1 {
  margin: 24px 0px 16px;
}

<div>
  <form
    class="c0"
    novalidate=""
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused Mui-required Mui-required"
        data-shrink="true"
        for="email"
        id="email-label"
      >
        Email Address
        <span
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          autocomplete="email"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="email"
          name="email"
          required=""
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-61 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-63 PrivateNotchedOutline-legendNotched-64"
          >
            <span>
              Email Address
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink="false"
        for="password"
        id="password-label"
      >
        Password
        <span
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          autocomplete="current-password"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-61 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-63"
          >
            <span>
              Password
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained c1 MuiButton-containedPrimary MuiButton-fullWidth"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Sign In
        </p>
      </span>
    </button>
  </form>
</div>
`;

exports[`LoginForm should show a generic error when one occurs during signing in 1`] = `
.c0 {
  width: 100%;
  margin-top: 8;
}

.c1.c1 {
  margin: 24px 0px 16px;
}

<form
  class="c0"
  novalidate=""
>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
      data-shrink="true"
      for="email"
      id="email-label"
    >
      Email Address
      <span
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        autocomplete="email"
        class="MuiInputBase-input MuiOutlinedInput-input"
        id="email"
        name="email"
        required=""
        type="text"
        value="test@test.com"
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-539 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-541 PrivateNotchedOutline-legendNotched-542"
        >
          <span>
            Email Address
             *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
      data-shrink="true"
      for="password"
      id="password-label"
    >
      Password
      <span
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        autocomplete="current-password"
        class="MuiInputBase-input MuiOutlinedInput-input"
        id="password"
        name="password"
        required=""
        type="password"
        value="somePassword"
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-539 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-541 PrivateNotchedOutline-legendNotched-542"
        >
          <span>
            Password
             *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <p
    class="MuiTypography-root MuiTypography-body2 MuiTypography-colorError"
  >
    Your username and/or password do not match
  </p>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained c1 MuiButton-containedPrimary MuiButton-fullWidth"
    tabindex="0"
    type="submit"
  >
    <span
      class="MuiButton-label"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Sign In
      </p>
    </span>
  </button>
</form>
`;

exports[`LoginForm should show an error when the user does not fill in username or password 1`] = `
.c0 {
  width: 100%;
  margin-top: 8;
}

.c1.c1 {
  margin: 24px 0px 16px;
}

<div>
  <form
    class="c0"
    novalidate=""
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-error Mui-error Mui-required Mui-required"
        data-shrink="false"
        for="email"
        id="email-label"
      >
        Email Address
        <span
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error Mui-error"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-describedby="email-helper-text"
          aria-invalid="true"
          autocomplete="email"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="email"
          name="email"
          required=""
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-187 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-189"
          >
            <span>
              Email Address
               *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained Mui-error Mui-required"
        id="email-helper-text"
      >
        Please enter your username
      </p>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-error Mui-error Mui-required Mui-required"
        data-shrink="false"
        for="password"
        id="password-label"
      >
        Password
        <span
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error Mui-error"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-describedby="password-helper-text"
          aria-invalid="true"
          autocomplete="current-password"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-187 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-189"
          >
            <span>
              Password
               *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained Mui-error Mui-required"
        id="password-helper-text"
      >
        Please enter your password
      </p>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained c1 MuiButton-containedPrimary MuiButton-fullWidth"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Sign In
        </p>
      </span>
    </button>
  </form>
</div>
`;

exports[`LoginForm should sign the user in the they enter username and password 1`] = `
.c0 {
  width: 100%;
  margin-top: 8;
}

.c1.c1 {
  margin: 24px 0px 16px;
}

<form
  class="c0"
  novalidate=""
>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
      data-shrink="true"
      for="email"
      id="email-label"
    >
      Email Address
      <span
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        autocomplete="email"
        class="MuiInputBase-input MuiOutlinedInput-input"
        id="email"
        name="email"
        required=""
        type="text"
        value="test@test.com"
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-363 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-365 PrivateNotchedOutline-legendNotched-366"
        >
          <span>
            Email Address
             *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
      data-shrink="true"
      for="password"
      id="password-label"
    >
      Password
      <span
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        autocomplete="current-password"
        class="MuiInputBase-input MuiOutlinedInput-input"
        id="password"
        name="password"
        required=""
        type="password"
        value="somePassword"
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-363 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-365 PrivateNotchedOutline-legendNotched-366"
        >
          <span>
            Password
             *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained c1 MuiButton-containedPrimary MuiButton-fullWidth"
    tabindex="0"
    type="submit"
  >
    <span
      class="MuiButton-label"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Sign In
      </p>
    </span>
  </button>
</form>
`;
